variant_caster cant handle c++ type nullptr_t

I have compile errors while compile this code (gcc 7.1 and vs 2017 checked)
```
    py::module m("xpytest", "pybind testing plugin");
    m.def("load_variant", [](std::variant<int, std::nullptr_t> v) { 
        return std::visit(visitor(), v);
    });
```
gcc output:
```
F:/devel/mingw-w64/i686-7.1.0-posix-dwarf-rt_v5-rev0/mingw32/bin/g++.exe  -c  "E:/code/zog/ptest3/xpytest.cpp" -O0 -g -std=c++17 -D_hypot=hypot -D_DEBUG  -o ./Debug/up_xpytest.cpp.o -I. -I. -If:\lang\Python35-32\include
In file included from f:\lang\Python35-32\include/pybind11/attr.h:13:0,
                 from f:\lang\Python35-32\include/pybind11/pybind11.h:36,
                 from E:/code/zog/ptest3/xpytest.cpp:3:
f:\lang\Python35-32\include/pybind11/cast.h: In instantiation of 'typename pybind11::detail::make_caster<T>::cast_op_type<T> pybind11::detail::cast_op(pybind11::detail::make_caster<T>&) [with T = std::nullptr_t; typename pybind11::detail::make_caster<T>::cast_op_type<T> = std::nullptr_t&; pybind11::detail::make_caster<T> = pybind11::detail::type_caster<std::nullptr_t>; typename pybind11::detail::intrinsic_type<T>::type = std::nullptr_t]':
f:\lang\Python35-32\include/pybind11/stl.h:305:31:   required from 'bool pybind11::detail::variant_caster<V<Ts ...> >::load_alternative(pybind11::handle, bool, pybind11::detail::type_list<U, Us ...>) [with U = std::nullptr_t; Us = {}; V = std::variant; Ts = {int, std::nullptr_t}]'
f:\lang\Python35-32\include/pybind11/stl.h:308:32:   required from 'bool pybind11::detail::variant_caster<V<Ts ...> >::load_alternative(pybind11::handle, bool, pybind11::detail::type_list<U, Us ...>) [with U = int; Us = {std::nullptr_t}; V = std::variant; Ts = {int, std::nullptr_t}]'
f:\lang\Python35-32\include/pybind11/stl.h:314:32:   required from 'bool pybind11::detail::variant_caster<V<Ts ...> >::load(pybind11::handle, bool) [with V = std::variant; Ts = {int, std::nullptr_t}]'
f:\lang\Python35-32\include/pybind11/cast.h:1469:9:   required from 'bool pybind11::detail::argument_loader<Args>::load_impl_sequence(pybind11::detail::function_call&, std::index_sequence<Indices1 ...>) [with unsigned int ...Is = {0}; Args = {std::variant<int, std::nullptr_t>}; std::index_sequence<Indices1 ...> = std::integer_sequence<unsigned int, 0>]'
f:\lang\Python35-32\include/pybind11/cast.h:1449:34:   required from 'bool pybind11::detail::argument_loader<Args>::load_args(pybind11::detail::function_call&) [with Args = {std::variant<int, std::nullptr_t>}]'
f:\lang\Python35-32\include/pybind11/pybind11.h:132:17:   required from 'pybind11::cpp_function::initialize(Func&&, Return (*)(Args ...), const Extra& ...)::<lambda(pybind11::detail::function_call&)> [with Func = pybind11_init()::<lambda(std::variant<int, std::nullptr_t>)>; Return = const char*; Args = {std::variant<int, std::nullptr_t>}; Extra = {pybind11::name, pybind11::scope, pybind11::sibling}]'
f:\lang\Python35-32\include/pybind11/pybind11.h:128:22:   required from 'struct pybind11::cpp_function::initialize(Func&&, Return (*)(Args ...), const Extra& ...) [with Func = pybind11_init()::<lambda(std::variant<int, std::nullptr_t>)>; Return = const char*; Args = {std::variant<int, std::nullptr_t>}; Extra = {pybind11::name, pybind11::scope, pybind11::sibling}]::<lambda(struct pybind11::detail::function_call&)>'
f:\lang\Python35-32\include/pybind11/pybind11.h:128:19:   required from 'void pybind11::cpp_function::initialize(Func&&, Return (*)(Args ...), const Extra& ...) [with Func = pybind11_init()::<lambda(std::variant<int, std::nullptr_t>)>; Return = const char*; Args = {std::variant<int, std::nullptr_t>}; Extra = {pybind11::name, pybind11::scope, pybind11::sibling}]'
f:\lang\Python35-32\include/pybind11/pybind11.h:62:9:   required from 'pybind11::cpp_function::cpp_function(Func&&, const Extra& ...) [with Func = pybind11_init()::<lambda(std::variant<int, std::nullptr_t>)>; Extra = {pybind11::name, pybind11::scope, pybind11::sibling}; <template-parameter-1-3> = void]'
f:\lang\Python35-32\include/pybind11/pybind11.h:743:22:   required from 'pybind11::module& pybind11::module::def(const char*, Func&&, const Extra& ...) [with Func = pybind11_init()::<lambda(std::variant<int, std::nullptr_t>)>; Extra = {}]'
```
msvc 2017 output:
```
1>------ Build started: Project: xpytest, Configuration: Debug Win32 ------
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.10.25017\bin\HostX86\x86\CL.exe /c /I"f:\lang\Python35-32\include" /I"C:\Program Files (x86)\Visual Leak Detector\include" /Zi /nologo /W3 /WX- /diagnostics:classic /Od /Ob0 /Oy- /D WIN32 /D _WINDOWS /D PYBIND11_CPP17 /D _DEBUG /D "CMAKE_INTDIR=\"Debug\"" /D xpytest_EXPORTS /D _WINDLL /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++latest /Fo"xpytest.dir\Debug\\" /Fd"xpytest.dir\Debug\vc141.pdb" /Gd /TP /analyze- /errorReport:prompt E:\code\zog\pytest2\xpytest.cpp
1>xpytest.cpp
1>f:\lang\python35-32\include\pybind11\common.h(30): warning C4005: 'PYBIND11_CPP17': macro redefinition
1>f:\lang\python35-32\include\pybind11\common.h(30): note: command-line arguments:  see previous definition of 'PYBIND11_CPP17'
1>f:\lang\Python35-32\include\pybind11/pybind11.h(148): warning C4101: 'dd': unreferenced local variable
1>f:\lang\Python35-32\include\pybind11/pybind11.h(63): note: see reference to function template instantiation 'void pybind11::cpp_function::initialize<_Ty,R,pybind11::handle,>(Func &&,Return (__cdecl *)(pybind11::handle))' being compiled
1>        with
1>        [
1>            _Ty=pybind11::detail::keep_alive_impl::<lambda_fcb9b197e266298ec55444104890fed0>,
1>            R=void,
1>            Func=pybind11::detail::keep_alive_impl::<lambda_fcb9b197e266298ec55444104890fed0>,
1>            Return=void
1>        ]
1>f:\lang\Python35-32\include\pybind11/pybind11.h(1310): note: see reference to function template instantiation 'pybind11::cpp_function::cpp_function<pybind11::detail::keep_alive_impl::<lambda_fcb9b197e266298ec55444104890fed0>,,void>(Func &&)' being compiled
1>        with
1>        [
1>            Func=pybind11::detail::keep_alive_impl::<lambda_fcb9b197e266298ec55444104890fed0>
1>        ]
1>f:\lang\python35-32\include\pybind11\cast.h(490): error C2738: 'operator std::nullptr_t &': is ambiguous or is not a member of 'pybind11::detail::type_caster<std::nullptr_t,void>'
1>f:\lang\python35-32\include\pybind11\cast.h(656): note: see declaration of 'pybind11::detail::type_caster<std::nullptr_t,void>'
1>f:\lang\Python35-32\include\pybind11/stl.h(305): note: see reference to function template instantiation '_Ty &pybind11::detail::cast_op<U>(pybind11::detail::type_caster<std::nullptr_t,void> &)' being compiled
1>        with
1>        [
1>            _Ty=std::nullptr_t,
1>            U=std::nullptr_t
1>        ]
1>f:\lang\Python35-32\include\pybind11/stl.h(308): note: see reference to function template instantiation 'bool pybind11::detail::variant_caster<std::variant<int,std::nullptr_t>>::load_alternative<std::nullptr_t,>(pybind11::handle,bool,pybind11::detail::type_list<std::nullptr_t>)' being compiled
1>f:\lang\Python35-32\include\pybind11/stl.h(308): note: see reference to function template instantiation 'bool pybind11::detail::variant_caster<std::variant<int,std::nullptr_t>>::load_alternative<std::nullptr_t,>(pybind11::handle,bool,pybind11::detail::type_list<std::nullptr_t>)' being compiled
1>f:\lang\Python35-32\include\pybind11/stl.h(314): note: see reference to function template instantiation 'bool pybind11::detail::variant_caster<std::variant<int,std::nullptr_t>>::load_alternative<int,std::nullptr_t>(pybind11::handle,bool,pybind11::detail::type_list<int,std::nullptr_t>)' being compiled
1>f:\lang\Python35-32\include\pybind11/stl.h(314): note: see reference to function template instantiation 'bool pybind11::detail::variant_caster<std::variant<int,std::nullptr_t>>::load_alternative<int,std::nullptr_t>(pybind11::handle,bool,pybind11::detail::type_list<int,std::nullptr_t>)' being compiled
1>f:\lang\Python35-32\include\pybind11/stl.h(313): note: while compiling class template member function 'bool pybind11::detail::variant_caster<std::variant<int,std::nullptr_t>>::load(pybind11::handle,bool)'
1>f:\lang\python35-32\include\pybind11\cast.h(1469): note: see reference to function template instantiation 'bool pybind11::detail::variant_caster<std::variant<int,std::nullptr_t>>::load(pybind11::handle,bool)' being compiled
1>f:\lang\Python35-32\include\pybind11/stl.h(329): note: see reference to class template instantiation 'pybind11::detail::variant_caster<std::variant<int,std::nullptr_t>>' being compiled
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.10.25017\include\tuple(199): note: see reference to class template instantiation 'pybind11::detail::type_caster<std::variant<int,std::nullptr_t>,void>' being compiled
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.10.25017\include\tuple(689): note: see reference to class template instantiation 'std::_Tuple_val<_This>' being compiled
1>        with
1>        [
1>            _This=pybind11::detail::type_caster<std::variant<int,std::nullptr_t>,void>
1>        ]
1>f:\lang\python35-32\include\pybind11\cast.h(1480): note: see reference to class template instantiation 'std::tuple<pybind11::detail::type_caster<std::variant<int,std::nullptr_t>,void>>' being compiled
1>f:\lang\Python35-32\include\pybind11/pybind11.h(124): note: see reference to class template instantiation 'pybind11::detail::argument_loader<std::variant<int,std::nullptr_t>>' being compiled
1>f:\lang\Python35-32\include\pybind11/pybind11.h(63): note: see reference to function template instantiation 'void pybind11::cpp_function::initialize<_Ty,R,std::variant<int,std::nullptr_t>,pybind11::name,pybind11::scope,pybind11::sibling>(Func &&,Return (__cdecl *)(std::variant<int,std::nullptr_t>),const pybind11::name &,const pybind11::scope &,const pybind11::sibling &)' being compiled
1>        with
1>        [
1>            _Ty=pybind11_init::<lambda_60b635692d52d37640d68fd3d08bd061>,
1>            R=const char *,
1>            Func=pybind11_init::<lambda_60b635692d52d37640d68fd3d08bd061>,
1>            Return=const char *
1>        ]
1>f:\lang\Python35-32\include\pybind11/pybind11.h(744): note: see reference to function template instantiation 'pybind11::cpp_function::cpp_function<_Ty,pybind11::name,pybind11::scope,pybind11::sibling,void>(Func &&,const pybind11::name &,const pybind11::scope &,const pybind11::sibling &)' being compiled
1>        with
1>        [
1>            _Ty=pybind11_init::<lambda_60b635692d52d37640d68fd3d08bd061>,
1>            Func=pybind11_init::<lambda_60b635692d52d37640d68fd3d08bd061>
1>        ]
1>E:\code\zog\pytest2\xpytest.cpp(26): note: see reference to function template instantiation 'pybind11::module &pybind11::module::def<pybind11_init::<lambda_60b635692d52d37640d68fd3d08bd061>,>(const char *,Func &&)' being compiled
1>        with
1>        [
1>            Func=pybind11_init::<lambda_60b635692d52d37640d68fd3d08bd061>
1>        ]
1>Done building project "xpytest.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 1 up-to-date, 0 skipped ==========
```

--- a/include/pybind11/cast.h
+++ b/include/pybind11/cast.h
@@ -653,7 +653,7 @@ private:
     void *value = nullptr;
 };
 
-template <> class type_caster<std::nullptr_t> : public type_caster<void_type> { };
+template <> class type_caster<std::nullptr_t> : public void_caster<std::nullptr_t> { };
 
 template <> class type_caster<bool> {
 public:

--- a/include/pybind11/cast.h
+++ b/include/pybind11/cast.h
@@ -602,7 +602,11 @@ public:
 
 template<typename T> struct void_caster {
 public:
-    bool load(handle, bool) { return false; }
+    bool load(handle src, bool) {
+        if (src && src.is_none())
+            return true;
+        return false;
+    }
     static handle cast(T, return_value_policy /* policy */, handle /* parent */) {
         return none().inc_ref();
     }

--- a/tests/test_python_types.cpp
+++ b/tests/test_python_types.cpp
@@ -516,6 +516,9 @@ test_initializer python_types([](py::module &m) {
             });
         }
     );
+
+    m.def("load_nullptr_t", [](std::nullptr_t) {}); // not useful, but it should still compile
+    m.def("cast_nullptr_t", []() -> std::nullptr_t { return {}; });
 });
 
 #if defined(_MSC_VER)

--- a/tests/test_python_types.cpp
+++ b/tests/test_python_types.cpp
@@ -359,9 +359,10 @@ test_initializer python_types([](py::module &m) {
         const char *operator()(int) { return "int"; }
         const char *operator()(std::string) { return "std::string"; }
         const char *operator()(double) { return "double"; }
+        const char *operator()(std::nullptr_t) { return "std::nullptr_t"; }
     };
 
-    m.def("load_variant", [](std::variant<int, std::string, double> v) {
+    m.def("load_variant", [](std::variant<int, std::string, double, std::nullptr_t> v) {
         return std::visit(visitor(), v);
     });
 
@@ -518,7 +519,7 @@ test_initializer python_types([](py::module &m) {
     );
 
     m.def("load_nullptr_t", [](std::nullptr_t) {}); // not useful, but it should still compile
-    m.def("cast_nullptr_t", []() -> std::nullptr_t { return {}; });
+    m.def("cast_nullptr_t", []() { return std::nullptr_t{}; });
 });
 
 #if defined(_MSC_VER)

--- a/tests/test_python_types.py
+++ b/tests/test_python_types.py
@@ -378,9 +378,10 @@ def test_variant(doc):
     assert load_variant(1) == "int"
     assert load_variant("1") == "std::string"
     assert load_variant(1.0) == "double"
+    assert load_variant(None) == "std::nullptr_t"
     assert cast_variant() == (5, "Hello")
 
-    assert doc(load_variant) == "load_variant(arg0: Union[int, str, float]) -> str"
+    assert doc(load_variant) == "load_variant(arg0: Union[int, str, float, None]) -> str"
 
 
 def test_constructors():
@@ -573,8 +574,5 @@ def test_capsule_with_destructor(capture):
 def test_void_caster():
     import pybind11_tests as m
 
-    with pytest.raises(TypeError) as excinfo:
-        m.load_nullptr_t(None)
-    assert "incompatible function arguments" in str(excinfo)
-
+    assert m.load_nullptr_t(None) is None
     assert m.cast_nullptr_t() is None

--- a/tests/test_python_types.py
+++ b/tests/test_python_types.py
@@ -568,3 +568,13 @@ def test_capsule_with_destructor(capture):
         creating capsule
         destructing capsule: 1234
     """
+
+
+def test_void_caster():
+    import pybind11_tests as m
+
+    with pytest.raises(TypeError) as excinfo:
+        m.load_nullptr_t(None)
+    assert "incompatible function arguments" in str(excinfo)
+
+    assert m.cast_nullptr_t() is None

